src/OpenSage.Game/Data/Apt/AptFile.cs

```diff
@@ -0,0 +1,88 @@
+﻿using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using OpenSage.Data.Apt.Characters;
+using OpenSage.Data.Utilities.Extensions;
+
+namespace OpenSage.Data.Apt
+{
+    public class AptFile
+    {
+        public ConstantData Constants { get; private set; }
+        public Movie Movie { get; private set; }
+        internal bool IsEmpty = true;
+        internal string MovieName;
+        internal FileSystem FileSystem;
+
+        public AptFile(ConstantData constants, FileSystem filesystem, string name)
+        {
+            Constants = constants;
+            FileSystem = filesystem;
+            MovieName = name;
+        }
+
+        public void Parse(BinaryReader reader)
+        {
+            //jump to the entry offset
+            var entryOffset = Constants.AptDataEntryOffset;
+            reader.BaseStream.Seek(entryOffset, SeekOrigin.Begin);
+
+            //proceed loading the characters
+            Movie = (Movie) Character.Create(reader, this);
+
+            //set first character to itself
+            Movie.Characters[0] = Movie;
+
+            var importDict = new Dictionary<string, AptFile>();
+
+            //resolve imports
+            foreach (var import in Movie.Imports)
+            {
+                //open the apt file where our character is located
+                AptFile importApt;
+
+                if (importDict.ContainsKey(import.Movie))
+                {
+                    importApt = importDict[import.Movie];
+                }
+                else
+                {
+                    var importEntry = FileSystem.GetFile(Path.ChangeExtension(import.Movie, ".apt"));
+                    importApt = AptFile.FromFileSystemEntry(importEntry);
```

Comment:
Presumably this could result in parsing an `.apt` file multiple times? That's why I was thinking this resolving of imports / exports would happen outside `AptFile`, in a separate layer that can make sure `.apt` files only get parsed once. What do you think?

Reply:
In theory that can happen, but parsing doesn't really require a lot of time. My idea was to have FromFileSystemEntry load an entire finished .apt, so you can't possibly end up with a crippled AptFile

Reply:
I like the goal of not being able to have an incomplete AptFile. Let's leave it as it is, but perhaps in the future we can pass `importDict ` into `AptFile.FromFileSystemEntry`, so they can be cached globally.